addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/handler/SamsungTvHandler.java

```diff
@@ -0,0 +1,343 @@
+			this.discoveryServiceRegistry.addDiscoveryListener(this);
+		}
+
+		if (upnpService != null) {
+			this.upnpService = upnpService;
+			this.upnpService.getRegistry().addListener(this);
+		} else {
+			logger.debug("upnpService not set.");
+		}
+
+		services = new ArrayList<>();
+	}
+
+	@Override
+	public void handleCommand(ChannelUID channelUID, Command command) {
+		logger.debug("Received channel: {}, command: {}", channelUID, command);
+
+		if (getThing().getStatus() == ThingStatus.ONLINE) {
+
+			// Delegate command to correct service
+
+			String channel = channelUID.getId();
+
+			for (SamsungTvService service : services) {
+				if (service != null) {
+					List<String> supportedCommands = service
+							.getSupportedChannelNames();
+					for (String s : supportedCommands) {
+						if (channel.equals(s)) {
+							service.handleCommand(channel, command);
+							return;
+						}
+					}
+				}
+			}
+
+			logger.warn("Channel '{}' not supported", channelUID);
+		} else {
+			logger.warn("Samsung TV '{}' is OFFLINE", getThing().getUID());
+		}
+	}
+
+	private synchronized void updatePowerState(boolean state) {
+		powerOn = state;
+	}
+
+	private synchronized boolean getPowerState() {
+		return powerOn;
+	}
+
+	/*
+	 * One Samsung TV contains several UPnP devices. Samsung TV is discovered by
+	 * Media Renderer UPnP device. This polling job tries to find another UPnP
+	 * devices related to same Samsung TV and create handler for those.
+	 */
+	private Runnable scanUPnPDevicesRunnable = new Runnable() {
+
+		@Override
+		public void run() {
+			logger.debug("Check UPnP devices");
+			checkAndCreateServices();
+		}
+	};
+
+	@Override
+	public void initialize() {
+		updateStatus(ThingStatus.OFFLINE);
+
+		configuration = getConfigAs(SamsungTvConfiguration.class);
+
+		logger.debug("Initializing Samsung TV handler for uid '{}'", getThing()
+				.getUID());
+
+		pollingJob = scheduler.schedule(scanUPnPDevicesRunnable, 0,
+				TimeUnit.MILLISECONDS);
+	}
+
+	@Override
+	public void dispose() {
+		if (pollingJob != null && !pollingJob.isCancelled()) {
+			pollingJob.cancel(true);
+			pollingJob = null;
+		}
+
+		stopServices();
+	}
+
+	@Override
+	public void thingDiscovered(DiscoveryService source, DiscoveryResult result) {
+		if (result.getThingUID().equals(this.getThing().getUID())) {
+			logger.debug("thingDiscovered");
+			if (configuration != null) {
+				updateStatus(ThingStatus.ONLINE);
+				updatePowerState(true);
+				updateState(new ChannelUID(getThing().getUID(), POWER),
+						OnOffType.ON);
+			} else {
+				logger.debug("thingDiscovered: Thing not yet initialized");
+			}
+		}
+	}
+
+	@Override
+	public void thingRemoved(DiscoveryService source, ThingUID thingUID) {
+		if (thingUID.equals(this.getThing().getUID())) {
+			logger.debug("thingRemoved: shutdown services");
+			updateStatus(ThingStatus.OFFLINE);
+			stopServices();
+			updateState(new ChannelUID(getThing().getUID(), POWER),
```

Comment:
This is very likely not correct to do here. The tv might vanish from UPnP because of some network issues, but it could still be powered on - so you should not imply here that it is off.